zio-http/src/main/scala/zio/http/Request.scala

```diff
@@ -133,43 +141,63 @@ object Request {
+        override final def context(implicit unsafe: Unsafe): ChannelHandlerContext = throw new IOException(
+          "Request does not have a context",
+        )
+        override final def encode(implicit unsafe: Unsafe): HttpRequest            = {
+          val jVersion = v.toJava
+          val path     = url.relative.encode
+          new DefaultFullHttpRequest(jVersion, method.toJava, path)
+        }
       }
-      override def body: Body                           = d
-      override def unsafeContext: ChannelHandlerContext = throw new IOException("Request does not have a context")
-
     }
   }
 
   private[zio] def fromFullHttpRequest(jReq: FullHttpRequest)(implicit ctx: Ctx): Request = {
 
     new Request {
-      override def method: Method            = Method.fromHttpMethod(jReq.method())
-      override def url: URL                  = URL.fromString(jReq.uri()).getOrElse(URL.empty)
-      override def headers: Headers          = Headers.make(jReq.headers())
-      override def body: Body                = Body.fromByteBuf(jReq.content())
-      override def version: Version          = Version.unsafeFromJava(jReq.protocolVersion())
-      override def unsafeEncode: HttpRequest = jReq
-      override def unsafeContext: Ctx        = ctx
+      override final def method: Method    = Method.fromHttpMethod(jReq.method())
+      override final def url: URL          = URL.fromString(jReq.uri()).getOrElse(URL.empty)
+      override final def headers: Headers  = Headers.make(jReq.headers())
+      override final def body: Body        = Body.fromByteBuf(jReq.content())
+      override final def version: Version  =
+        Unsafe.unsafe { implicit u =>
+          Version.unsafe.fromJava(jReq.protocolVersion())
+        }
+      override final val unsafe: UnsafeAPI = new UnsafeAPI {
+        override final def encode(implicit unsafe: Unsafe): HttpRequest = jReq
+        override final def context(implicit unsafe: Unsafe): Ctx        = ctx
+      }
     }
   }
 
   private[zio] def fromHttpRequest(jReq: HttpRequest)(implicit ctx: Ctx): Request = {
     new Request {
-      override def headers: Headers          = Headers.make(jReq.headers())
-      override def method: Method            = Method.fromHttpMethod(jReq.method())
-      override def url: URL                  = URL.fromString(jReq.uri()).getOrElse(URL.empty)
-      override def version: Version          = Version.unsafeFromJava(jReq.protocolVersion())
-      override def unsafeEncode: HttpRequest = jReq
-      override def unsafeContext: Ctx        = ctx
-      override def body: Body = Body.fromAsync { async => ServerInboundHandler.Unsafe.addAsyncBodyHandler(async) }
+      override final def headers: Headers = Headers.make(jReq.headers())
+      override final def method: Method   = Method.fromHttpMethod(jReq.method())
+      override final def url: URL         = URL.fromString(jReq.uri()).getOrElse(URL.empty)
+      override final def version: Version =
```

Comment:
😬 Not sure we really want `Unsafe` here. Ideally we would have "safe by construction" rather than "safe by deconstruction", to push the problem onto constructor.